Better error message when specializing generic abstract type with unit

smoothdeveloper opened the pull request:
Attempt to fix #35

I'll need some guidance in terms of where to add tests and of course if there is anything wrong with the matching I'm doing / assumptions I'm taking.

I'm concerned the matching might yield false positive if there are cases where using `unit` as a generic type argument would be valid, but this is difficult for me to reason about as of now.

Also looking for better ideas for the error message.

forki commented:
regarding tests: tests\fsharpqa\Source\Warnings\env.lst is now a dumping ground for tests that check error messages. I think it would be ok to add it there

forki commented on src/fsharp/CompileOps.fs:
```diff
@@ -1144,15 +1145,32 @@ let OutputPhasedErrorR (os:System.Text.StringBuilder) (err:PhasedError) =
+                  | _ :: ts -> hasUnitTType_app ts
+                  | [] -> false
+
+              match minfoVirt.EnclosingType with
```
most simmilar pattern matches call striptypEqns - maybe @dsyme can tell if it's needed here

smoothdeveloper commented:
thanks @forki, I'll pay a look and try to add few tests supposed to trigger that new message.

smoothdeveloper commented:
Mmmh reading details @dsyme left in the ancient conversation and my proposed fix is not really handling things at the level he was intending the changes to happen but later down the road, although it looks simpler (if that doesn't raise false positives), maybe it doesn't follow the normal conventions for deciding of errors, but there are few other conditionals to decide which error message to raise where I've added the logic.

> To add a better error message, the logical place to put it would be here: https://github.com/fsharp/fsharp/blob/6ac6318858da25fe4f1d4728901c5824e9f3bf3a/src/fsharp/typrelns.fs#L1191

> "IsExactMatch" would need to be duplicated and adjusted to a "IsExactMatchUpToInstantiatedUnitReturnType". If an override exists satisfying that predicate, then a better error message can then be given, much as for code already present to give a better error message when there is an override satisfying IsPartialMatch or IsNameMatch.

I'm trying to see if other cases would trigger the initial / new error messages, this compiles fine with and without the change:

```fsharp
type Foo<'t> =
  abstract member Bar : 't -> int

type Bar() =
  interface Foo<unit> with
    member x.Bar _ = 1
```

I'll look if fsharpqa also has tests checking no error / warning shows up to add that case.

If anyone knows of additional cases I should account for, that would be great.

enricosada commented on tests/fsharpqa/Source/Conformance/TypesAndTypeConstraints/TypeParameterDefinitions/UnitSpecialization.fs:
```diff
@@ -0,0 +1,9 @@
+// #UnitGenericAbstractType 
+//<Expects status=success></Expects>
```
it's not an error (this sintax is supported), but can you pls write that as xml like (supported too)? 

```
//<Expects status="success"></Expects>
```

i am trying to convert the suite to nunit, and to simplify i'll need to convert that as xml like (easier to parse, validate)

enricosada commented on tests/fsharpqa/Source/ErrorMessages/UnitGenericAbstractType/E_UnitGenericAbstractType1.fs:
```diff
@@ -0,0 +1,9 @@
+// #ErrorMessages #UnitGenericAbstractType 
+//<Expects status="error" span="(7,21)" id="FS0017">The member 'Apply : int -> unit' is specialized with 'unit' but 'unit' can't be used as return type of an abstract method parameterized on return type.\.</Expects>
```
the regex at end `.\.` is right? is not just `\.`?
like that parse any character followed by dot, instead of just a dot

But i just tried with this https://regex101.com/r/oG5sD1/1 online regex, if you remove the `.` before `\.` works.

smoothdeveloper replied:
@enricosada thanks!

I need more time to dig into TESTGUIDE.md but it seems there is a lot of implicit knowledge about how those tests are defined which is undocumented.

I tried to follow stuff around (the wrong xml stuff was copied from another file) without knowing exactly how that works.

smoothdeveloper commented:
@KevinRansom or @dsyme could we get

smoothdeveloper commented:
@KevinRansom or @dsyme could we get a review on proposed fix and some direction for the error message?

KevinRansom commented:
@dsyme 
This looks good to me, but I may be missing something, could you take a look please.

Kevin

KevinRansom commented:
Thank you for taking care of this

Kevin
